presto-main/src/main/java/com/facebook/presto/operator/aggregation/arrayagg/SetUnionFunction.java

```diff
@@ -0,0 +1,162 @@
+import static com.facebook.presto.operator.aggregation.AggregationMetadata.ParameterMetadata.ParameterType.NULLABLE_BLOCK_INPUT_CHANNEL;
+import static com.facebook.presto.operator.aggregation.AggregationMetadata.ParameterMetadata.ParameterType.STATE;
+import static com.facebook.presto.spi.function.Signature.typeVariable;
+import static com.facebook.presto.util.Reflection.methodHandle;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+public class SetUnionFunction
+        extends SqlAggregationFunction
+{
+    public static final SetUnionFunction SET_UNION = new SetUnionFunction();
+    private static final String NAME = "set_union";
+    private static final MethodHandle INPUT_FUNCTION = methodHandle(SetUnionFunction.class, "input", Type.class, Type.class, SetAggregationState.class, Block.class, int.class);
+    private static final MethodHandle COMBINE_FUNCTION = methodHandle(SetUnionFunction.class, "combine", SetAggregationState.class, SetAggregationState.class);
+    private static final MethodHandle OUTPUT_FUNCTION = methodHandle(SetUnionFunction.class, "output", SetAggregationState.class, BlockBuilder.class);
+
+    public SetUnionFunction()
+    {
+        super(NAME,
+                ImmutableList.of(typeVariable("T")),
+                ImmutableList.of(),
+                parseTypeSignature("array(T)"),
+                ImmutableList.of(parseTypeSignature("array(T)")));
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "return an array of unique values";
+    }
+
+    @Override
+    public InternalAggregationFunction specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionManager functionManager)
+    {
+        Type type = boundVariables.getTypeVariable("T");
+        ArrayType outputType = (ArrayType) typeManager.getParameterizedType(StandardTypes.ARRAY, ImmutableList.of(
+                TypeSignatureParameter.of(type.getTypeSignature())));
+        return generateAggregation(type, outputType);
+    }
+
+    private static InternalAggregationFunction generateAggregation(Type type, ArrayType outputType)
+    {
+        DynamicClassLoader classLoader = new DynamicClassLoader(SetUnionFunction.class.getClassLoader());
+
+        List<Type> inputTypes = ImmutableList.of(type);
+        AccumulatorStateSerializer<?> stateSerializer = new SetAggregationStateSerializer(outputType);
+        AccumulatorStateFactory<?> stateFactory = new SetAggregationStateFactory(type);
+
+        Type intermediateType = stateSerializer.getSerializedType();
+        List<ParameterMetadata> inputParameterMetadata = createInputParameterMetadata(type);
+        Class<? extends AccumulatorState> stateInterface = SetAggregationState.class;
+        AggregationMetadata metadata = new AggregationMetadata(
+                AggregationUtils.generateAggregationName(NAME, outputType.getTypeSignature(), inputTypes.stream().map(Type::getTypeSignature).collect(toImmutableList())),
```

Comment:
It's probably more efficient to just do `ImmutableList.of(type.getTypeSignature())` here rather than creating an `ImmutableList` of one element and iterate through it.

Reply:
Done